client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -85,25 +116,28 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                curTotal -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None
+                  if curTotal < maxTotal && allocated.getOrElse(key, 0) < getMaxConnections(key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if idleQueue.get(key).exists(_.nonEmpty) =>
                 logger.debug(
                   s"No connections available for the desired key. Evicting oldest and creating a new connection: $stats")
-                allocated -= 1
-                idleQueue.dequeue().shutdown()
+                curTotal -= 1
+                decrConnection(key)
```

Comment:
Oh, I meant for `curTotal` to be part of those helper functions.  Because you're always doing them together.